File: zig/ui/glfw/shader.zig
const std = @import("std");
const AutoHashMap = std.AutoHashMap;
const gl = @cImport(@cInclude("GL/glew.h"));

const ShaderType = enum(usize) {
    VERTEX = 0,
    FRAGMENT = 1,
};

const ShaderProgramSource = struct {
    VertexSource: []u8 = undefined,
    FragmentSource: []u8 = undefined,
};


fn parseShader(filepath: []const u8, allocator: std.mem.Allocator) !ShaderProgramSource {
    var source = [2]std.ArrayList(u8){try std.ArrayList(u8).initCapacity(allocator, 0), try std.ArrayList(u8).initCapacity(allocator, 0)};
    var courceType: ShaderType = .VERTEX;
    var prog: ShaderProgramSource = .{};
    const f = try std.fs.cwd().openFile(filepath, .{.mode = .read_only});
    defer f.close();
    var f_buf: [4096]u8 = undefined;
    var f_reader = f.reader(&f_buf);

    while (f_reader.interface.takeDelimiterExclusive('\n')) |line| {
        if (std.mem.indexOf(u8, line, "#shader")) |_| {
            if (std.mem.indexOf(u8, line, "vertex")) |_| {
                courceType = .VERTEX;
            }
            if (std.mem.indexOf(u8, line, "fragment")) |_| {
                courceType = .FRAGMENT;
            }
        } else {
            try source[@intFromEnum(courceType)].appendSlice(allocator, line);
            try source[@intFromEnum(courceType)].append(allocator, '\n');
        }
    } else |_| {
    }

    prog.VertexSource = try allocator.dupe(u8, source[@intFromEnum(ShaderType.VERTEX)].items);
    prog.FragmentSource = try allocator.dupe(u8, source[@intFromEnum(ShaderType.FRAGMENT)].items);

    source[@intFromEnum(ShaderType.VERTEX)].deinit(allocator);
    source[@intFromEnum(ShaderType.FRAGMENT)].deinit(allocator);

    return prog;
}


fn compileShader(source: []const u8, _type: u32, allocator: std.mem.Allocator) !u32 {
    const id = gl.glCreateShader().?(_type);
    gl.glShaderSource().?(id, 1, @ptrCast(&source), null);
    gl.glCompileShader().?(id);

    var res: c_int = 0;
    gl.glGetShaderiv().?(id, gl.GL_COMPILE_STATUS, &res);
    if (res == gl.GL_FALSE) {
        var length: c_int = 0;
        gl.glGetShaderiv().?(id, gl.GL_INFO_LOG_LENGTH, &length);
        const message: []u8 = try allocator.alloc(u8, @intCast(length));
        gl.glGetShaderInfoLog().?(id, length, &length, message.ptr);
        std.debug.print("Failed to compile {s} shader {s}\n", .{
            if (_type == gl.GL_VERTEX_SHADER) "vertex" else "fragment",
            message
        });
        gl.glDeleteShader().?(id);
        return 0;
    }
    return id;
}

fn createShader(vertexShader: []const u8, fragmentShader: []const u8, allocator: std.mem.Allocator) !u32 {
    const program = gl.glCreateProgram().?();
    std.debug.print("Programm: {d}\n", .{@as(u32, program)});
    const vs: u32 = try compileShader(vertexShader, gl.GL_VERTEX_SHADER, allocator);
    const fs: u32 = try compileShader(fragmentShader, gl.GL_FRAGMENT_SHADER, allocator);

    gl.glAttachShader().?(program, vs);
    gl.glAttachShader().?(program, fs);
    gl.glLinkProgram().?(program);

    gl.glDeleteShader().?(vs);
    gl.glDeleteShader().?(fs);

    return @as(u32, @intCast(program));
}

pub fn create(filepath: []const u8, allocator: std.mem.Allocator) !Shader {
    var s = Shader{.filepath = filepath};
    const prog_source: ShaderProgramSource = try parseShader(filepath, allocator);
    std.debug.print("Vertex Shader:\n{s}\n", .{prog_source.VertexSource});
    std.debug.print("Fragment Shader:\n{s}\n", .{prog_source.FragmentSource});

    defer {
        allocator.free(prog_source.VertexSource);
        allocator.free(prog_source.FragmentSource);
    }

    s.rendererID = try createShader(prog_source.VertexSource, prog_source.FragmentSource, allocator);
    s.uniform_location_catch = std.AutoHashMap([]u8, u32).init(allocator);

    return s;
}

pub const Shader = struct {
    filepath: []const u8,
    rendererID: u32 = 0,
    uniform_location_catch: AutoHashMap([]u8, u32) = undefined,
    allocator: std.mem.Allocator = undefined,

    pub fn destroy(self: *Shader) void {
        gl.glDeleteProgram().?(self.rendererID);
        self.uniform_location_catch.deinit();
    }

    pub fn bind(self: *Shader) void {
        gl.glUseProgram().?(self.rendererID);

    }

    pub fn unbind(self: *Shader) void {
        _ = self;
        gl.glUseProgram().?(0);
    }

    fn getUniformLocation(self: *Shader, name: []const u8) i32 {
        if (self.uniform_location_catch.contains(name)) {
            return self.uniform_location_catch.get(name).?;
        }

        const location: i32 = gl.glGetUniformLocation().?(self.rendererID, name.ptr);
        if (location == -1) {
            std.debug.print("Cant find shader {s} variable {s} location", .{self.filepath, name});
            return -1;
        }

        self.uniform_location_catch.put(name, @as(u32, @intCast(location)));

        return location;
    }

    pub fn setUniform4f(self: *Shader, name: []const u8, v0: f32, v1: f32, v2: f32, v3: f32) void {
        gl.glUniform4f().?(self.getUniformLocation(name), v0, v1, v2, v3);
    }
};

File: c/glfw/ch1/Shader.c
#include "Shader.h"
#include <GL/glew.h>
#include <stdio.h>
#include <string.h>


static ShaderProgramSource ShaderParseShader(str file) {
  reader *r = reader_create_from_file(file);
  str_buf *sb[2] = {str_buf_create(), str_buf_create()};

  u32 type;
  ShaderType st = NONE;
  
  str line;
  while((line = reader_read_until_del(r, '\n')) != null) {
    if (strstr(line, "#shader") != null) {
      if (strstr(line, "vertex") != null) {
	st = VERTEX;
      } else if (strstr(line, "fragment") != null) {
	st = FRAGMENT;
      }
    } else {
      str_buf_append_format(sb[st], "%s\n", line);
    }
  }
  ShaderProgramSource prog = {.VertexSource = str_buf_to_string(sb[0]), .FragmentSource = str_buf_to_string(sb[1])};
  str_buf_destroy(sb[0]);
  str_buf_destroy(sb[1]);

  return prog;
}

static u32 ShaderCompileShader(const str source, u32 type) {
  u32 id = glCreateShader(type);
  GLCall(glShaderSource(id, 1, (const char *const *)&source, null));
  GLCall(glCompileShader(id));

  i32 result;
  GLCall(glGetShaderiv(id, GL_COMPILE_STATUS, &result));
  if (result == GL_FALSE) {
    i32 length;
    GLCall(glGetShaderiv(id, GL_INFO_LOG_LENGTH, &length));
    byte *message = (byte *)alloca(length + 1);
    GLCall(glGetShaderInfoLog(id, length, &length, message));
    log(ERROR, "Failed compile %s shader %s", type == GL_VERTEX_SHADER ? "vertex" : "fragment", message);
    GLCall(glDeleteShader(id));
    return 0;
  }
  
  return id;
}

static u32 ShaderCreateShader(const str vertexShader, const str fragmentShader) {
  u32 program = glCreateProgram();
  u32 vs = ShaderCompileShader(vertexShader, GL_VERTEX_SHADER);
  u32 fs = ShaderCompileShader(fragmentShader, GL_FRAGMENT_SHADER);

  GLCall(glAttachShader(program, vs));
  GLCall(glAttachShader(program, fs));
  GLCall(glLinkProgram(program));
  GLCall(glValidateProgram(program));

  GLCall(glDeleteShader(vs));
  GLCall(glDeleteShader(fs));
  
  return program;
}


Shader ShaderCreate(char *filepath) {
    Shader s = {.filepath = filepath};

    ShaderProgramSource source = ShaderParseShader(filepath);
    s.rendererID = ShaderCreateShader(source.VertexSource, source.FragmentSource);
    s.uniformLocationCache = map_create(STR, I32, null, null);

    return s;
}
void ShaderDestroy(Shader *s) {
    GLCall(glDeleteProgram(s->rendererID));
    map_destroy(s->uniformLocationCache);
}

void ShaderBind(Shader *s) {
    GLCall(glUseProgram(s->rendererID));
}

void ShaderUnbind(Shader *s) {
    GLCall(glUseProgram(0));
}

static i32 ShaderGetUniformLocation(Shader *s, const char *name) {
    if (map_contain(s->uniformLocationCache, (ptr)name)) {
        return *(i32 *)map_get(s->uniformLocationCache, (ptr)name).val;
    }

    i32 location = 0;
    GLCall(location = glGetUniformLocation(s->rendererID, name));
    if (location == -1) {
        log(ERROR, "Cant find shader %s variable %s location", s->filepath, name);
        return -1;
    }

    map_put(s->uniformLocationCache, (ptr)name, &location);

    return location;
}

void ShaderSetUniform4f(Shader *s, const char *name, float v0, float v1, float v2, float v3) {
    GLCall(glUniform4f(ShaderGetUniformLocation(s, name), v0, v1, v2, v3));

}

void ShaderSetUniform1i(Shader *s, const char *name, u32 v0) {
    GLCall(glUniform1i(ShaderGetUniformLocation(s, name), v0));
}
